app/services/pii/encryptor.rb

```diff
@@ -0,0 +1,65 @@
+module Pii
+  class Encryptor
+    DELIMITER = '.'.freeze
+    DIGEST = OpenSSL::Digest::SHA256.new.freeze
+
+    # "It is a riddle, wrapped in a mystery, inside an enigma; but perhaps there is a key."
+    #  - Winston Churchill, https://en.wiktionary.org/wiki/a_riddle_wrapped_up_in_an_enigma
+    #
+
+    def initialize(key_maker = Pii::KeyMaker.new)
+      @key_maker = key_maker
+      @cipher = Pii::Cipher.new
+    end
+
+    def encrypt(plaintext, cek)
+      payload = sign_and_concat(plaintext)
+      encode(cipher.encrypt(payload, cek))
+    end
+
+    def decrypt(ciphertext, cek)
+      decrypt_and_test_payload(decode(ciphertext), cek)
+    end
+
+    def sign(text)
+      signing_key = key_maker.signing_key
+      encode(signing_key.sign(DIGEST, encode(text)))
+    end
+
+    private
+
+    attr_reader :key_maker, :cipher
+
+    def sign_and_concat(plaintext)
+      plaintext_signature = sign(plaintext)
+      join_segments(plaintext, plaintext_signature)
+    end
+
+    def decrypt_and_test_payload(payload, cek)
+      payload = cipher.decrypt(payload, cek)
+      raise EncryptionError unless sane_payload?(payload)
+      plaintext, plaintext_signature = split_into_segments(payload)
+      return plaintext if sign(plaintext) == plaintext_signature
```

Comment:
I'm not sure I understand the logic here, signature verification should not be resigning and testing that it matches, it should go through the `verify` API.